cpp/server/handler.cc

```diff
@@ -237,38 +241,81 @@ int64_t GetIntParam(const multimap<string, string>& query,
 }  // namespace
 
 
-HttpHandler::HttpHandler(LogLookup<LoggedCertificate>* log_lookup,
-                         const ReadOnlyDatabase<LoggedCertificate>* db,
-                         const CertChecker* cert_checker, Frontend* frontend,
-                         ThreadPool* pool)
+HttpHandler::HttpHandler(
+    LogLookup<LoggedCertificate>* log_lookup,
+    const ReadOnlyDatabase<LoggedCertificate>* db,
+    const ClusterStateController<LoggedCertificate>* controller,
+    const CertChecker* cert_checker, Frontend* frontend, Proxy* proxy,
+    ThreadPool* pool)
     : log_lookup_(CHECK_NOTNULL(log_lookup)),
       db_(CHECK_NOTNULL(db)),
+      controller_(CHECK_NOTNULL(controller)),
       cert_checker_(CHECK_NOTNULL(cert_checker)),
       frontend_(frontend),
+      proxy_(CHECK_NOTNULL(proxy)),
       pool_(CHECK_NOTNULL(pool)) {
 }
 
 
+void HttpHandler::ProxyInterceptor(
+    const libevent::HttpServer::HandlerCallback& local_handler,
+    evhttp_request* request) {
+  // Need to punt this via the threadpool because the interaction with the
+  // controller can block pending other libevent updates, and since we're on
+  // the libevent thread here...
+  pool_->Add([this, request, local_handler]() {
+    VLOG(1) << "Running proxy interceptor...";
+    // TODO(alcutter): We can be a bit smarter about when to proxy off the
+    // request - being stale wrt to the current serving STH doesn't
+    // automatically mean we're unable to answer this request.
+    if (controller_->NodeIsStale()) {
+      const util::Status status(proxy_->ProxyRequest(request));
+      if (!status.ok()) {
+        LOG(INFO) << "Failed to proxy request: " << status;
+        SendError(request, HTTP_INTERNAL, "Failed to proxy request");
+      }
+      return;
+    } else {
+      local_handler(request);
```

Comment:
Uh oh... This is no-go, as the handler is then going to (indirectly) call ```evhttp_send_reply```, which I'm pretty sure needs to be on the libevent thread.

I'm thinking that the check of ```NodeIsStale``` could be done before spinning this off?

This could be a little simpler (just something like ```pool_->Add(bind(&Proxy::ProxyRequest, proxy_, request));```) if ```ProxyRequest``` would take over the request *completely* (and send the error if there's one).